src/commands/run.rs

```diff
@@ -524,6 +589,72 @@ impl RunCommand {
     }
 }
 
+struct GuestProfiler {
+    profile: Profile,
+    process: fxprof_processed_profile::ProcessHandle,
+    thread: fxprof_processed_profile::ThreadHandle,
+    start: Instant,
+}
+
+impl GuestProfiler {
+    fn new(interval: Duration) -> Self {
+        let mut profile = Profile::new(
+            "Wasmtime",
+            std::time::SystemTime::now().into(),
+            interval.into(),
+        );
+        let process = profile.add_process("main", 0, Timestamp::from_nanos_since_reference(0));
+        let thread = profile.add_thread(process, 0, Timestamp::from_nanos_since_reference(0), true);
+        let start = Instant::now();
+        Self {
+            profile,
+            process,
+            thread,
+            start,
+        }
+    }
+
+    fn sample(&mut self, store: impl AsContext) {
+        let now = Timestamp::from_nanos_since_reference(
+            (Instant::now() - self.start).as_nanos().try_into().unwrap(),
+        );
+
+        let trace = WasmBacktrace::force_capture(store);
+        // Samply needs to see the oldest frame first, but we list the newest
+        // first, so iterate in reverse.
+        let frames = Vec::from_iter(trace.frames().iter().rev().map(|frame| {
+            let frame = if let Some(name) = frame.func_name() {
+                let idx = self.profile.intern_string(name);
+                Frame::Label(idx)
+            } else {
+                // `func_offset` should always be set because we force
+                // `generate_address_map` on if profiling is enabled.
+                let func_offset = frame.func_offset().unwrap();
+                Frame::InstructionPointer(func_offset.try_into().unwrap())
+            };
+            FrameInfo {
+                frame,
+                category_pair: CategoryHandle::OTHER.into(),
+                flags: FrameFlags::empty(),
+            }
```

Comment:
Are there any flags or configuration we could perhaps pass to indicate which module a frame comes from? Or perhaps which instance a frame comes from? We don't currently have a means of learning that from a wasm frame but it's something we could theoretically add.